driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java

```diff
@@ -568,7 +568,11 @@ private static void refreshNodeListAndTokenMap(Connection connection, Cluster.Ma
                 // We don't know that node, create the Host object but wait until we've set the known
                 // info before signaling the addition.
                 host = cluster.metadata.add(foundHosts.get(i));
-                isNew = true;
+                if (host == null)
+                    // There must have been a race condition and someone else added the node before us
+                    host = cluster.metadata.getHost(foundHosts.get(i));
```

Comment:
This is not atomic with the last call to add. I know it's highly unlikely, but in theory host could be null again after this line.

We could split `Metadata.add` into two methods:
* `Metadata.newHost(InetSocketAddress)` that creates the Host object
* `Metadata.addIfAbsent(Host)` with putIfAbsent semantics (returns the existing Host if it already existed)

Then we could write this:
```java
// We don't know that node, create the Host object but wait until we've set the known
// info before signaling the addition.
Host newHost = cluster.metadata.newHost(foundHosts.get(i));
Host existing = cluster.metadata.addIfAbsent(newHost);
if (existing == null) {
    host = newHost;
    isNew = true;
} else {
    host = existing;
    isNew = false;
}
```
Then we need to address the other places where `add` was used, but that's easy.